components/script/dom/htmlslotelement.rs

```diff
@@ -356,6 +357,32 @@ impl HTMLSlotElement {
         debug_assert!(self.is_in_agents_signal_slots.get());
         self.is_in_agents_signal_slots.set(false);
     }
+
+    /// Returns the slot's assigned nodes if the root's slot assignment mode
+    /// is "named", or the manually assigned nodes otherwise
+    pub(crate) fn assigned_nodes(&self) -> Ref<'_, [Slottable]> {
+        Ref::map(self.assigned_nodes.borrow(), Vec::as_slice)
+    }
+}
+
+macro_rules! match_slottable {
+    (
+        $slottable: expr,
+        $element_fn: expr,
+        $text_fn: expr,
+    ) => {
+        match $slottable.inner.type_id() {
+            NodeTypeId::Element(_) => {
+                let element: &Element = $slottable.inner.downcast::<Element>().unwrap();
+                $element_fn(element)
+            },
+            NodeTypeId::CharacterData(CharacterDataTypeId::Text(_)) => {
+                let text: &Text = $slottable.inner.downcast::<Text>().unwrap();
+                $text_fn(text)
+            },
+            _ => unreachable!(),
+        }
+    };
```

Comment:
Perhaps this could be a method on `Slottable` instead since you are passing anonymous functions anyway. That way you could avoid metaprogramming.

Reply:
I think I initially didn't do that because of ownership semantics (using a function doesn't work when there's a non-`Copy` variable used in both branches, because both closures would need ownership of it). I'll see if I can make it work.